mart/attack/threat_model.py

```diff
@@ -4,43 +4,122 @@
 # SPDX-License-Identifier: BSD-3-Clause
 #
 
+from __future__ import annotations
+
 import abc
-from typing import Any, Dict, Union
+from typing import Any
 
 import torch
 
-__all__ = ["BatchThreatModel"]
+__all__ = ["BatchComposer"]
+
+
+class Constraint(abc.ABC):
+    @abc.abstractclassmethod
+    def __call__(self, input, target, input_adv) -> None:
+        raise NotImplementedError
+
+
+class Range(Constraint):
+    def __init__(self, min, max):
+        self.min = min
+        self.max = max
+
+    def __call__(self, input, target, input_adv):
+        if torch.any(input_adv < self.min) or torch.any(input_adv > self.max):
+            raise ValueError(f"Adversarial input is outside [{self.min}, {self.max}].")
+
+
+class Lp(Constraint):
+    def __init__(self, eps: float, p: int | float | None = torch.inf, dim=None, keepdim=False):
+        self.p = p
+        self.eps = eps
+        self.dim = dim
+        self.keepdim = keepdim
+
+    def __call__(self, input, target, input_adv):
+        perturbation = input_adv - input
+        norm_vals = perturbation.norm(p=self.p, dim=self.dim, keepdim=self.keepdim)
+        norm_max = norm_vals.max()
+        if norm_max > self.eps:
+            raise ValueError(
+                f"L-{self.p} norm of perturbation exceeds {self.eps}, reaching {norm_max}"
+            )
+
+
+class Integer(Constraint):
+    def __call__(self, input, target, input_adv):
+        torch.testing._comparison.assert_equal(input_adv, input_adv.round())
+
+
+class Mask(Constraint):
+    def __call__(self, input, target, input_adv):
+        # True/1 is mutable, False/0 is immutable.
+        # mask.shape=(H, W)
+        mask = target["perturbable_mask"]
+
+        # Immutable boolean mask, True is immutable.
+        imt_mask = (1 - mask).bool()
+        perturbation = input_adv - input
+        if perturbation.masked_select(imt_mask).any():
```

Comment:
If this is slow (which is might be because masked select would seem like a slow operation), you could probably get away with using Lp(eps=0) on mask*input_adv, mask*input.

Reply:
Hmm okay.

Reply:
I don't quite understand but there is probably some subtly I'm missing.

Reply:
You are right. I have simplified the implementation.